Add documentation for Namespaces and Interfaces (#531)
kadena-io/pact: docs/en/pact-reference.md

admin authorization schemes for modules and tables. Definition creates the keysets
in the runtime environment and stores their definition in the global keyset database.

#### Namespace declaration {#namespacedefinition}

> [Namespace](#namespaces) declarations provide a unique prefix for code defined within the namespace scope. All smart contract definitions must occur within a namespaced scope, which is initiated by defining a namespace using [define-namespace](#define-namespace), and entering the namespace scope by issuing the [namespace](#namespace) command. The global (i.e. empty) namespace is reserved for Kadena code.

Reviewer: ```suggestion
[Namespace](#namespaces) declarations provide a unique prefix for modules and interfaces defined within the namespace scope. Namespaces are handled differently in public and private blockchain contexts: in private they are freely definable, and the _root namespace_ (ie, not using a namespace at all) is available for user code. In public blockchains, users are not allowed to use the root namespace (which is reserved for built-in contracts like the coin contract) and must define code within a namespace, which may or may not be definable (ie, users might be restricted to "user" namespaces). 

Namespaces are defined using [define-namespace](#define-namespace). Namespaces are "entered" by issuing the [namespace](#namespace) command. 
```

When a module is declared, all references to native functions or definitions from other modules are resolved. Resolution failure results in transaction rollback.

Modules can be re-defined as controlled by their admin keyset. Module versioning is not supported, except by including a version sigil in the module name (e.g., "accounts-v1"). However, *module hashes* are a powerful feature for ensuring code safety. When a module is imported with [use](#use), the module hash can be specified, to tie code to a particular release.

As of Pact 2.2, `use` statements can be issued within a module declaration. This combined with module hashes provides a high level of assurance, as updated module code will fail to import if a dependent module has subsequently changed on the chain; this will also propagate changes to the loaded modules' hash, protecting downstream modules from inadvertent changes on update.

Module names must be globally unique.

#### Interface Declaration {#interfacedeclaration}

[Interfaces](#interface) contain an API specification and data definitions for smart contracts.
They are comprised of:

- [function](#defun) specifications (i.e. function signatures)

> - [constant](#defconst) values

Reviewer 2: ```suggestion
- [constant](#defconst) values
- formal verification models
```
Add explicit mention of models

Author: If we add this here, we should add it to the modules and defun section as well. Deferring to @slpopejoy for a y/n on that - there may be reasoning I don't know about.

Author: Added

admin authorization schemes for modules and tables. Definition creates the keysets
in the runtime environment and stores their definition in the global keyset database.

#### Namespace declaration {#namespacedefinition}

> [Namespace](#namespaces) declarations provide a unique prefix for keysets and modules defined within the namespace scope. All smart contract definitions must occur within a namespaced scope, which is initiated by defining a namespace using [define-namespace](pact-functions.html#define-namespace), and entering the namespace scope by issuing the [namespace](pact-functions.html#namespace) command. The global (i.e. empty) namespace is reserved for Kadena code.

Reviewer: ```suggestion
[Namespace](#namespaces) declarations provide a unique prefix for modules defined within the namespace scope. All smart contract definitions must occur within a namespaced scope, which is initiated by defining a namespace using [define-namespace](#define-namespace), and entering the namespace scope by issuing the [namespace](#namespace) command. The global (i.e. empty) namespace is reserved for Kadena code.
```

Reviewer: Also get rid of "The global ..." and make a para on Chainweb namespaces vs private blockchain (where namespaces are fine).

Chainweb is a WIP.

#### Module declaration {#moduledeclaration}

> [Modules](#module) contain an API implementation and data definitions for smart contracts.

Reviewer: Can we leave lines alone that have no changes? Git isn't indicating what changed so I would rather you revert unchanged paras.

```

### Namespaces {#namespaces}

Namespaces are [defined](pact-functions.html#define-namespace) by specifying a namespace name and [associating](pact-functions.html#read-keyset)
a keyset with the namespace. Namespace scope is entered by [declaring](pact-functions.html#namespace) the namespace environment. All definitions issued after the namespace scope is entered will be accessible by their fully qualified names. These names are of the form _namespace.module.definition_. This form can also be used to access code outside of the current namespace for the purpose of importing module code, or implementing modules:

```lisp
(implements my-namespace.my-interface)
;; or
(use my-namespace.my-module)
```

> Code may be appended to the namespace by simply entering the re-entering the namespace and declaring new code definitions. All definitions _must_ occur within a namespace, as the global namespace (the empty namespace) is reserved for Kadena code.

Author: We are not _currently_ enforcing this, no. However, we will eventually, and the intent is to have this be a thing by the time we release if not very soon after.

Examples of valid namespace definition and scoping:

#### Example: Defining a namespace

Defining a namespace requires a keyset, and a namespace name of type string:

```lisp
(begin-tx)
(env-data { "my-keyset" : ["my-keys"] })
(env-keys ["my-keys"])

(define-keyset 'my-keyset)
(define-namespace 'my-namespace (read-keyset 'my-keyset))
(commit-tx)

pact> (namespace 'my-namespace)
"Namespace set to my-namespace"
```

#### Example: Accessing members of a namespace

> Members of a namespace my be access by their fully-qualified names:

Author: Thanks linda!

```lisp
(begin-tx)

(namespace 'my-namespace)

(module my-module EXAMPLE_GUARD

  (defcap EXAMPLE_GUARD ()
    (enforce-keyset 'my-keyset))

  (defun hello-number:string (number:integer)
    (format "Hello, your number is {}!" [number]))
)
(commit-tx)

pact> (use my-namespace.my-module)
"Using my-namespace.my-module"
pact> (hello-number 3)
"Hello, your number is 3!"
```

#### Example: Importing module code or implementing interfaces at a namespace

> Modules my be imported at a namespace, and interfaces my be implemented in a similar way. This allows the user to work with members of a namespace in a much less verbose and cumbersome way.

Reviewer 2: ```suggestion
Modules may be imported at a namespace, and interfaces my be implemented in a similar way. This allows the user to work with members of a namespace in a much less verbose and cumbersome way.
```

```lisp
(begin-tx)

(env-data { "my-keyset" : ["foo"] })
(env-keys ["foo"])

(define-keyset 'my-keyset)

(define-namespace 'my-namespace (read-keyset 'my-keyset))
(define-namespace 'my-other-namespace (read-keyset 'my-keyset))

(commit-tx)

(begin-tx)

(namespace 'my-namespace)

(interface my-interface

  (defun hello-number:string (number:integer))
)

(commit-tx)
(begin-tx)

(namespace 'my-other-namespace)
(module my-module EXAMPLE_GUARD

  (implements my-namespace.my-interface)

  (defcap EXAMPLE_GUARD ()
    (enforce-keyset 'my-keyset))

  (defun hello-number:string (number:integer)
    (format "Hello, your number is {}!" [number]))
)

(commit-tx)

pact> (use my-other-namespace.my-module)
"Using my-other-namespace.my-module"
pact> (hello-number 3)
"Hello, your number is 3!"

```

> However, if one is simply appending code to an existing namespace, then the namespace prefix in the fully qualified name may be ommitted, as using a namespace works in a similar way to importing a module: all toplevel definitions within a namespace are brought into scope when `(namespace 'my-namespace)` is declared.

Reviewer 2: Quick example?

Author: I could come up with one, sure

When models are declared in an interface, they are appeneded to the list of models present in the implementing module at the level of declaration: body-level models are appended to body-level models, and function-level models are appended to function-level models. This allows users to extend the constraints of an interface with models applicable to specific business logic and implementation.

Declaring models shares the same syntax with modules:

#### Example: declaring models, tables, and importing modules in an interface

```lisp
(begin-tx)

> (env-data { "admin-keyset": ["product-dept" "dev-dept"] })

Reviewer 2: ```suggestion
```
not needed for this example

(module acct-module ACCT-GOVERNANCE
  @doc "account schema module"

  (defcap ACCT-GOVERNANCE ()
    true)

  (defschema account
    balance:integer
    ks:keyset
    active:bool
  )
)
(commit-tx)
(begin-tx)

(interface coin-sig

  "Coin Contract Abstract Interface Example"

  (use acct-module)

  (defun transfer:string (from:string to:string amount:integer)
    @doc   "Transfer money between accounts"
    @model [(property (row-enforced accounts "ks" from))
            (property (> amount 0))

>             (property (= 0 (column-delta accounts "balance")))

Reviewer 2: Referring to:
```lisp
(column-delta accounts "balance")
```
This is tricky because `accounts` table is created in another module that implements this interface and not in `acct-module`. Not sure if this _should_ be allowed, but maybe we can just add the following to `acct-module` to avoid any confusion:
```lisp
(deftable accounts:{account})
```

Author: Good call. Added.

```

> ### Namespaces {#namespaces}

Reviewer: Why is this here and not above (or the above below down here?) Cf. guards and capabilities which are one section each/in one place.

Namespaces are [defined](pact-functions.html#define-namespace) by specifying a namespace name and [associating](pact-functions.html#read-keyset)
a keyset with the namespace. Namespace scope is entered by [declaring](pact-functions.html#namespace) the namespace environment. All definitions issued after the namespace scope is entered will be accessible by their fully qualified names. These names are of the form _namespace.module.definition_. This form can also be used to access code outside of the current namespace for the purpose of importing module code, or implementing modules:

```lisp
(implements my-namespace.my-interface)
;; or
(use my-namespace.my-module)
```

Code may be appended to the namespace by simply entering the re-entering the namespace and declaring new code definitions. All definitions _must_ occur within a namespace, as the global namespace (the empty namespace) is reserved for Kadena code.

Examples of valid namespace definition and scoping:

#### Example: Defining a namespace

Defining a namespace requires a keyset, and a namespace name of type string:

> ```lisp

Reviewer: It's not the practice to give full-on repl examples. However I'd like to keep these around so maybe add to `pact-examples` repo? @LindaOrtega can give you details.

Reviewer: ```
pact> (define-namespace 'my-namespace (read-keyset 'my-keyset))
"[defined message]"
pact> (namespace 'my-namespace)
"Namespace set to my-namespace"
```

Author: Fixed. I'll come up with some examples for `pact-examples` that are less simplistic
